jlm/llvm/opt/alias-analyses/RegionAwareModRefSummarizer.cpp

```diff
@@ -256,275 +201,132 @@ class RegionSummary final
   {
-    return { RegionSummaryConstIterator(RegionSummaries_.begin()),
-             RegionSummaryConstIterator(RegionSummaries_.end()) };
+    return NodeMap_.find(&node) != NodeMap_.end();
   }
 
-  [[nodiscard]] RegionSummary *
-  TryGetRegionSummary(const rvsdg::Region & region) const
+  [[nodiscard]] ModRefSetIndex
+  GetSetForNode(const rvsdg::Node & node) const
   {
-    const auto it = RegionSummaries_.find(&region);
-    if (it == RegionSummaries_.end())
-      return nullptr;
-    return it->second.get();
+    const auto it = NodeMap_.find(&node);
+    JLM_ASSERT(it != NodeMap_.end());
+    return it->second;
   }
 
-  [[nodiscard]] bool
-  ContainsRegionSummary(const rvsdg::Region & region) const
+  [[nodiscard]] ModRefSetIndex
+  GetOrCreateSetForNode(const rvsdg::Node & node)
   {
-    return TryGetRegionSummary(region) != nullptr;
-  }
+    if (const auto it = NodeMap_.find(&node); it != NodeMap_.end())
+      return it->second;
 
-  [[nodiscard]] RegionSummary &
-  GetRegionSummary(const rvsdg::Region & region) const
-  {
-    const auto regionSummary = TryGetRegionSummary(region);
-    JLM_ASSERT(regionSummary != nullptr);
-    return *regionSummary;
+    return NodeMap_[&node] = CreateModRefSet();
   }
 
-  RegionSummary &
-  AddRegionSummary(std::unique_ptr<RegionSummary> regionSummary)
+  void
+  MapNodeToSet(const rvsdg::Node & node, ModRefSetIndex index)
   {
-    JLM_ASSERT(!ContainsRegionSummary(regionSummary->GetRegion()));
-
-    auto region = &regionSummary->GetRegion();
-    auto regionSummaryPointer = regionSummary.get();
-    RegionSummaries_[region] = std::move(regionSummary);
-    return *regionSummaryPointer;
+    JLM_ASSERT(!HasSetForNode(node));
+    NodeMap_[&node] = index;
   }
 
-  [[nodiscard]] CallSummaryRange
-  GetCallSummaries()
+  const util::HashSet<const PointsToGraph::MemoryNode *> &
+  GetSimpleNodeModRef(const rvsdg::SimpleNode & node) const override
   {
-    return { CallSummaryIterator(CallSummaries_.begin()),
-             CallSummaryIterator(CallSummaries_.end()) };
+    return ModRefSets_[GetSetForNode(node)].GetMemoryNodes();
   }
 
-  [[nodiscard]] CallSummaryConstRange
-  GetCallSummaries() const
+  const util::HashSet<const PointsToGraph::MemoryNode *> &
+  GetGammaEntryModRef(const rvsdg::GammaNode & gamma) const override
   {
-    return { CallSummaryConstIterator(CallSummaries_.begin()),
-             CallSummaryConstIterator(CallSummaries_.end()) };
+    return ModRefSets_[GetSetForNode(gamma)].GetMemoryNodes();
   }
 
-  [[nodiscard]] CallSummary *
-  TryGetCallSummary(const rvsdg::SimpleNode & call) const
+  const util::HashSet<const PointsToGraph::MemoryNode *> &
+  GetGammaExitModRef(const rvsdg::GammaNode & gamma) const override
   {
-    const auto it = CallSummaries_.find(&call);
-    if (it == CallSummaries_.end())
-      return nullptr;
-    return it->second.get();
+    return ModRefSets_[GetSetForNode(gamma)].GetMemoryNodes();
```

Comment:
Could just call GetGammaEntryModeRef() here. This would make it clear that a gamma always has the same entry/exit memory nodes.